vm/ubpf_vm.c

```diff
@@ -577,20 +581,62 @@ int
 ubpf_run(struct ubpf_vm *vm, void *mem, size_t mem_len, bool jit,
     uint64_t *bpf_return_value, char **errmsg)
 {
+    int cntrfd = 0;
+    ubpf_jit_fn fn;
+
+    if (vm->inst_cnt) {
+        cntrfd = setup_instruction_counter();
+        if (cntrfd == -1) {
+            *errmsg = ubpf_error("Unable to setup instruction counter");
+            return -1;
+        }
+    }
+
     if (jit) {
-        ubpf_jit_fn fn = ubpf_compile(vm, errmsg);
+        if (vm->inst_cnt) {
+            enable_instruction_count(cntrfd);
+        }
+        fn = ubpf_compile(vm, errmsg);
+        if (vm->inst_cnt) {
+            disable_instruction_count(cntrfd);
+            vm->inst_cnt->inst_cmpl = get_instruction_count(cntrfd);
+            if (vm->inst_cnt->inst_cmpl < 0) {
+                *errmsg = ubpf_error("Unable to setup instruction counter");
+                return -1;
+            }
+        }
         if (fn == NULL) {
             return -1;
         }
+    }
+
+    if (vm->inst_cnt) {
+        enable_instruction_count(cntrfd);
+    }
+    if (jit) {
         *bpf_return_value = fn(mem, mem_len);
     } else {
         if (ubpf_exec(vm, mem, mem_len, bpf_return_value) < 0) {
             *bpf_return_value = UINT64_MAX;
         }
     }
+    if (vm->inst_cnt) {
+        disable_instruction_count(cntrfd);
+        vm->inst_cnt->inst = get_instruction_count(cntrfd);
+        if (vm->inst_cnt->inst < 0) {
+            *errmsg = ubpf_error("Unable to get instruction count");
+            return -1;
+        }
+    }
     return 0;
 }
 
+void
+ubpf_set_inst_cnt(struct ubpf_vm *vm, struct ubpf_inst_cnt *inst_cnt)
```

Comment:
@iomartin this can't be static as it is a prototype for the library and is exposed to the outside world. I don't want to do it as an inline since there are no inlines in there yet. So I will leave this as is.